Tools/ardupilotwaf/ap_library.py

```diff
@@ -0,0 +1,197 @@
+        _vehicle_indexes[vehicle] = len(_vehicle_indexes) + 1
+    return _vehicle_indexes[vehicle]
+
+_vehicle_macros = ('SKETCHNAME', 'SKETCH', 'APM_BUILD_DIRECTORY',
+                   'APM_BUILD_TYPE')
+_macros_re = re.compile(r'\b(%s)\b' % '|'.join(_vehicle_macros))
+
+def _remove_comments(s):
+    return c_preproc.re_cpp.sub(c_preproc.repl, s)
+
+_depends_on_vehicle_cache = {}
+def _depends_on_vehicle(bld, source_node):
+    path = source_node.srcpath()
+
+    if path not in _depends_on_vehicle_cache:
+        s = _remove_comments(source_node.read())
+        _depends_on_vehicle_cache[path] = _macros_re.search(s) is not None
+
+    return _depends_on_vehicle_cache[path]
+
+@conf
+def ap_library(bld, library, vehicle):
+    try:
+        common_tg = bld.get_tgen_by_name(_common_tgen_name(library))
+    except Errors.WafError:
+        common_tg = None
+
+    try:
+        vehicle_tg = bld.get_tgen_by_name(_vehicle_tgen_name(library, vehicle))
+    except Errors.WafError:
+        vehicle_tg = None
+
+    if common_tg and vehicle_tg:
+        return
+
+    library_dir = bld.srcnode.find_dir('libraries/%s' % library)
+    if not library_dir:
+        bld.fatal('ap_library: %s not found' % library)
+
+    src = library_dir.ant_glob(ap.SOURCE_EXTS + UTILITY_SOURCE_EXTS)
+
+    if not common_tg:
+        kw = dict(bld.env.AP_LIBRARIES_OBJECTS_KW)
+        kw['features'] = kw.get('features', []) + ['ap_library_object']
+        kw.update(
+            name=_common_tgen_name(library),
+            source=[s for s in src if not _depends_on_vehicle(bld, s)],
+            idx=0,
+        )
+        bld.objects(**kw)
+
+    if not vehicle_tg:
+        source = [s for s in src if _depends_on_vehicle(bld, s)]
+
+        if not source:
+            return
+
+        kw = dict(bld.env.AP_LIBRARIES_OBJECTS_KW)
+        kw['features'] = kw.get('features', []) + ['ap_library_object']
+        kw.update(
+            name=_vehicle_tgen_name(library, vehicle),
+            source=source,
+            defines=ap.get_legacy_defines(vehicle),
+            idx=_vehicle_index(vehicle),
+        )
+        bld.objects(**kw)
+
+@before_method('process_use')
+@feature('cxxstlib')
+def process_ap_libraries(self):
+    self.use = Utils.to_list(getattr(self, 'use', []))
+    libraries = Utils.to_list(getattr(self, 'ap_libraries', []))
+    vehicle = getattr(self, 'ap_vehicle', None)
+
+    for l in libraries:
+        self.use.append(_common_tgen_name(l))
+        if vehicle:
+            self.use.append(_vehicle_tgen_name(l, vehicle))
+
+class ap_library_check_header(Task.Task):
+    color = 'PINK'
+    tasks = []
+
+    def run(self):
+        self.tasks.append(self)
+        n = self.inputs[0]
+        s = _remove_comments(n.read())
+        self.guilty = _macros_re.search(s) is not None
+
+    def post_run(self):
+        super(ap_library_check_header, self).post_run()
+        if self.guilty:
+            self.generator.bld.task_sigs[self.uid()] = None
```

Comment:
Force task to run again in the next check if the header was invalid. Otherwise the header wouldn't be checked again and silently fool the developer. The way this is implemented is the same way done for `waf_unit_tests` tool.